Access transmission frames

iiacoban42 opened the pull request:
Adjusted the db schema:  
- TimeFields are now DateTimeFields to also include the date
- the frame field is now a TextField to be able to add hex data as string but I added a frame_binary column for the binary if we're gonna use binary at any point. 
- Added a radio_amateur field to Downlink table for submitted telemetry
- Da Vinci table is the same as Delfi with the addition of command and content codes.

Added a query and filter functionality to the Uplink/Donwlink tables + a script that adds a lists of json frames to the Downlink table, still have to fix the timestamp for this one. What's the exact timestamp format that we'll use? Same as the dummy data or something else?

Closes #18
Closes #19   
Closes #40 
Closes #21

StefanoSperetta commented:
The timestamp format was quite random in my example, especially because it comes from an existing application that produces a text file and I turned it into JSON with a piece of Java. I would prefer a "standard" JSON timestamp (which does not really exist formally but I would go for **2012-04-23T18:25:43.511Z** as in here: https://stackoverflow.com/questions/10286204/what-is-the-right-json-date-format) because it allows to use directly a timestamp object in the code / database. I can re-generate the dummy data with any type of timestamp you want, so we make our life easy.

Thanks for the reset_docker script, I was looking for it myself. Would it be possible to add a mention in the README file?

I haven't had time yet to look fully at the code (hopefully tonight) but what is the standard entry point? I see a script in the ewigs folder so I guess this would only be the entry for "our" GS but for the other amateurs? I have no very hard preference but it might seem easier to have a unified submission of data. From the ewigs code, it seems all frames submitted by our GS end up in an ewi gs database, then a processing script splits them per satellite?

iiacoban42 commented:
Alright, I'll update the timestamp format and readme. `add_downlink_frames` from `src/ewilgs/views.py` could be used internally as well as for other amateurs. With `get_downlink/uplink_frames` the frames are retrieved and could also be filtered by any combination of attributes. I'll need to make a few tweaks to `add_downlink_frames` though:
- turn it into a POST request 
- make it login protected to avoid spam.
- support file uploads (for amateurs). What's the common practice when sharing telemetry? CSV files, JSON or something else?
- add the load more frames button and a GUI for the filtering menu.

I'll add these in a follow-up PR if that's okay, this one is getting a bit big.

StefanoSperetta commented:
The most common system is using dedicated applications that demodulate and submit frames one by one. The submission method is highly custom (everybody has its own method, as an example https://github.com/daniestevez/gr-satellites/blob/master/python/bme_submitter.py). I want to submit a dedicated submitter application for gr-satellites as well, it is quite simple. Usually nobody uploads files but prefers real-time uploads.

iiacoban42 commented:
Great to hear, that's much better than handling files!

iiacoban42 commented:
Should be all set! I updated the timestamp format and it uses UTC everywhere. Also refactored the filtering script and added a gui for it, now we're filtering the django way. The results are paginated, the limit is set to 100 rows for now. You should be able to see the downlink table at `http://127.0.0.1:8000/ewilgs/downlink/` after adding the data to the db (only need to request `http://127.0.0.1:8000/ewilgs/downlink/addframe` and check the comments in `add_downlink_frames`). Let me know if you'd like me to make any other changes.

StefanoSperetta commented:
From what I see (but I have not tested the code yet), all submitted frames end up in a table in the ewigs app (this sis my guess from the path of the Uplink and Downlink models): I did not consider this solution before but I have no issue with that, it basically means our ewigs app is the mage gateway to the system. Based on this, would it be possible to rename it to a more general name and have a simpler and more intuitive URL? 

Looking at the url.py, I think you might have a very simple path like /submit (relative to the main app) to make it easier to understand (or variations of this).

NotUnlikeTheWaves commented on reset_docker.sh:
```diff
@@ -0,0 +1,6 @@
+# resets and removes the volumes from a docker container
+docker-compose down
+docker rm -f $(docker ps -a -q)
+docker volume rm $(docker volume ls -q)
```
This seems like it would remove every docker container and volume, not just the ones related to this project. Can you think of a way to adjust it so that it only removes the DelifTLM instances? Aside from that very nice to have this in a script

NotUnlikeTheWaves replied:
`docker-compose down -v` might be an option here: https://docs.docker.com/compose/reference/down/

iiacoban42 replied:
Thanks, that's a good one!

NotUnlikeTheWaves commented on src/delfic3/migrations/0001_initial.py:
```diff
@@ -1,4 +1,4 @@
-# Generated by Django 3.2.8 on 2021-12-12 19:55
```
I would expect this to generate a new migration on top of the initial, did you remove them locally before or does django work in this way?

NotUnlikeTheWaves replied:
Ah, I see there is no content change in this file so that's all good ;). I'm still curious why it regenerated this file

iiacoban42 replied:
I did run into some issues and then I removed the migration files and recreated them :). Django doesn't modify existing migration files

NotUnlikeTheWaves commented on src/davinci/migrations/0001_initial.py:
```diff
@@ -1,4 +1,4 @@
-# Generated by Django 3.2.8 on 2021-12-12 19:55
+# Generated by Django 3.2.8 on 2021-12-19 19:20
```
Maybe my comment is more applicable here: I would expect the changes in this file to come from a new migration script instead of amending it to the current one, since django keeps track of which migrations it has performed by the app name and file name alone. There is a risk with an existing DB that the changes here will not be applied to it because it has already executed the script name before

StefanoSperetta reviewed (approved):
I agree with @iiacoban42, we can move ahead with this PR and fix the remaining stuff in the next one. What would be important for me is handling POST requests, login and looking for a more general name for the app (or simply registering a simpler submit path).

iiacoban42 commented:
Changed the url to `/submit`, so that's done.

NotUnlikeTheWaves commented on reset_docker.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+# restarts and removes the volumes from a docker container
+docker-compose down -v
+docker-compose up -d
```
Maybe one last thing (can be a next PR too) is to add the --build flag for outdated images (I find it doesn't automatically build it after changes)
